_posts/2022-9-5-Diplomat @ Google.md

```diff
@@ -0,0 +1,503 @@
+how it works, but details on the motivation can be found in the [design document](https://github.com/rust-diplomat/diplomat/blob/main/docs/design_doc.md).
+
+Diplomat is a unidirectional FFI tool that allows programs in other languages
+to call Rust libraries via automatically generated, idiomatic APIs. At its core
+lays the abstract representation of types and methods, which represents the
+lowest common denominator for what types and methods look like across most
+modern programming languages. Since the abstract representation is so general,
+Diplomat is theoretically capable of generating bindings to any language that
+Rust has FFI support for, namely JavaScript via WebAssembly, and languages that
+support C FFI. Currently, it has support for generating bindings to C, C++, C#,
+and JavaScript/TypeScript.  Diplomat also generates native-feeling types and
+methods in the target language that call into the raw WASM/C interface.
+
+The cost of creating bindings to all these languages is just one set of FFI
+bindings, which are exposed through a module marked with the
+`#[diplomat::bridge]` proc macro. So if you have some large Rust library like
+ICU4X, you can attach Diplomat bindings and get native feeling APIs in every
+supported language at no extra cost.
+
+# Abstract Representation
+
+In general, there are two kinds of structs in programming. The first kind of
+struct has values that are semantically disjoint, where fields can typically be
+mutated without breaking things. In Rust, the most common examples are tuples,
+but it also shows up in config types frequently. The second kind are opaque
+structs which usually have some more advanced functionality. An example is
+Rust's `HashMap`, where the user shouldn't particulary care about the exact
+inner fields, and only cares about the functionality.
+
+Diplomat distinguishes between these two kinds of structs, and the programmer
+can communicate this by marking opaque structs with the `#[diplomat::opaque]`
+attribute. Structs marked as opaque have special properties in Diplomat's model:
+1. Fields are private from the other language, only methods can be used.
+2. Fields can be _any_ valid Rust values, since Diplomat doesn't examine them.
+3. Can only cross the FFI boundary behind a pointer, meaning there is no
+conversion cost.
+4. Boxed opaques can only be returned, but not accepted as a parameters because
+most other languages don't have a way to give up ownership.
+
+On the other hand, structs not marked with the `#[diplomat::opaque]` attribute
+have nearly opposite properties:
+1. Fields are public to the other language.
+2. Fields can only be primitive types like numbers, booleans, characters,
+owned opaques, and other structs. My project extended this to also allow for
+borrowed opaques, string slices, and slices of primitives.
+3. Can only cross the FFI boundary by value, meaning that every field is also
+converted at the boundary.
+
```

Comment:
I think it's worth mentioning what happens to these two types of structs in JS as an example

(opaque: JS stores a pointer into wasm memory, and has a destructor that cleans it up. non-opaque: treated identically to a JS `{....}` object with each field having the right type)